stl/inc/xstring

```diff
@@ -2526,38 +2467,78 @@ private:
         _Apply_annotation(_My_data._Myptr(), _My_data._Myres, _My_data._Mysize, _My_data._Myres);
     }
 
-    _CONSTEXPR20 void _Modify_annotation(const difference_type _Count) const noexcept {
-        // Extends/shrinks the annotated range by _Count
-        if (_Count == 0) {
+    _CONSTEXPR20 void _Modify_annotation(const size_type _Old_size, const size_type _New_size) const noexcept {
+        if (_Old_size == _New_size) {
             return;
         }
 
         auto& _My_data = _Mypair._Myval2;
-        _Apply_annotation(
-            _My_data._Myptr(), _My_data._Myres, _My_data._Mysize, static_cast<size_type>(_My_data._Mysize + _Count));
+        _Apply_annotation(_My_data._Myptr(), _My_data._Myres, _Old_size, _New_size);
     }
 
-    _NODISCARD static const void* _Get_aligned_first(const void* _First, const size_type _Capacity) noexcept {
-        const char* _CFirst = reinterpret_cast<const char*>(_First);
-
-        if (_Capacity >= _Asan_granularity) { // We are guaranteed to have sufficient space to find an aligned address
-            return reinterpret_cast<const void*>(
-                (reinterpret_cast<uintptr_t>(_CFirst) + (_Asan_granularity - 1)) & ~(_Asan_granularity - 1));
+    _CONSTEXPR20 static void _Remove_sso_annotation(const _Scary_val& _My_data) noexcept {
+        // Removes annotation of the SSO buffer with shadow memory.
+        // This results in the ability to copy into the SSO buffer without worrying.

```

Comment:
I imagine the extra calls here are mostly about removing the reliance on the string's current state for `_Create_annotation` and `_Remove_annotation` to make sure we're always doing what we intend? Can they be implemented in terms of `_Apply_annotation` to reduce code duplication?

Reply:
They cannot be; these are functions on the SSO buffer _regardless_ of the current SSO state of the string.

Reply:
_Apply_annotation is static though - it doesn't use any container state, just what's passed in. _Remove_sso_annotation could just be:
```
_Apply_annotation(_My_data._Bx._Buf, _BUF_SIZE - 1, 0, _BUF_SIZE - 1);
```
No?

That said, I see the clarity value in using `__sanitizer_annotate_contiguous_container` directly. It's mostly the complication that comes with having multiple code paths that bothers me and needing to worry about how those interweave.

Reply:
I'm essentially just worried that some SSO buffer annotations are removed with `_Remove_sso_annotation` and some are removed with `_Apply_annotation`. It does appear to be fine in practice though since `_Apply_annotation` is the one with the potentially smaller brush and it is also the only one that poisons, so I'm convinced it doesn't need to change.